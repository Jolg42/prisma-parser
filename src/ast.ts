export type Node =
  | DocumentNode
  | DefinitionNode
  | FieldDefinitionNode
  | TypeNode
  | IdentifierNode
  | AttributeNode
  | ConfigOptionNode
  | ExpressionNode

export type Position = {
  line: number
  column: number
  offset: number
}

export type Location = {
  start: Position
  end: Position
}

export type DocumentNode = {
  kind: 'Document'
  definitions: DefinitionNode[]
  location?: Location
}

export type DefinitionNode = ModelDefinitionNode | ConfigDefinitionNode

export type ConfigDefinitionNode = {
  kind: 'ConfigDefinition'
  type: ConfigType
  name: IdentifierNode
  options: ConfigOptionNode[]
  location?: Location
}

export type ConfigType = 'datasource' | 'generator'

export type ConfigOptionNode = {
  kind: 'ConfigOption'
  key: IdentifierNode
  value: ExpressionNode
  location?: Location
}

export type ExpressionNode = StringLiteralNode | BooleanLiteralNode | FunctionCallNode

export type StringLiteralNode = {
  kind: 'StringLiteral'
  value: string
  location?: Location
}

export type BooleanLiteralNode = {
  kind: 'BooleanLiteral'
  value: boolean
  location?: Location
}

export type FunctionCallNode = {
  kind: 'FunctionCall'
  name: IdentifierNode
  arguments: ExpressionNode[]
  location?: Location
}

export type ModelDefinitionNode = {
  kind: 'ModelDefinition'
  name: IdentifierNode
  fields: FieldDefinitionNode[]
  location?: Location
}

export type FieldDefinitionNode = {
  kind: 'FieldDefinition'
  name: IdentifierNode
  type: TypeNode
  attributes: AttributeNode[]
  location?: Location
}

export type TypeModifier = 'none' | 'array' | 'optional'

export type TypeNode = {
  kind: 'Type'
  name: IdentifierNode
  modifier: TypeModifier
  location?: Location
}

export type IdentifierNode = {
  kind: 'Identifier'
  identifier: string
  location?: Location
}

export type AttributeNode = {
  kind: 'Attribute'
  name: string
  location?: Location
}
