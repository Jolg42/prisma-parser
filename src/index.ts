export { parse } from './parser'
export { print } from './printer'
export { PrismaParsingError } from './prisma-parsing-error'
export * from './ast'
