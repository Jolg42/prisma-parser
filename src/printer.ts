import {
  DocumentNode,
  FieldDefinitionNode,
  Node,
  TypeNode,
  AttributeNode,
  DefinitionNode,
  ModelDefinitionNode,
  ConfigDefinitionNode,
  ConfigOptionNode,
  ExpressionNode,
  BooleanLiteralNode,
  StringLiteralNode,
  FunctionCallNode,
} from './ast'
import { assertNever } from './util'

const TAB_SIZE = 2

type PrintFunction<NodeType extends Node> = (node: NodeType, printerState: PrinterState) => void

function padBy(paddingSize: number, str: string): string {
  return str.padStart(str.length + paddingSize)
}

class PrinterState {
  private indentationLevel = 0
  private lines: string[] = []
  private currentLine = ''

  indent(): this {
    this.indentationLevel++
    return this
  }

  unindent(): this {
    if (this.indentationLevel === 0) {
      throw new Error('Can not unindent beyond 0')
    }
    this.indentationLevel--
    return this
  }

  write(string: string): this {
    this.currentLine += string
    return this
  }

  writeAligned(string: string, offsetFromStartOfTheLine: number): this {
    const paddingSize = Math.max(offsetFromStartOfTheLine - this.currentLine.length, 0)
    this.write(padBy(paddingSize, string))
    return this
  }

  newLine(): this {
    this.lines.push(padBy(this.indentationLevel * TAB_SIZE, this.currentLine))
    this.currentLine = ''
    return this
  }

  writeSeparated<NodeType extends Node>(
    nodes: NodeType[],
    separator: string,
    printNode: PrintFunction<NodeType>,
  ): this {
    if (nodes.length === 0) {
      return this
    }
    printNode(nodes[0], this)
    for (let i = 1; i < nodes.length; i++) {
      if (separator === '\n') {
        this.newLine()
      } else {
        this.write(separator)
      }
      printNode(nodes[i], this)
    }
    return this
  }

  getText(): string {
    if (this.currentLine) {
      this.newLine()
    }
    return this.lines.join('\n')
  }
}

export function print(document: DocumentNode): string {
  const printerState = new PrinterState()
  printDocument(document, printerState)
  return printerState.getText()
}

function printDocument(document: DocumentNode, printerState: PrinterState): void {
  printerState.writeSeparated(document.definitions, '\n', printDefinition)
  printerState.newLine()
}

function printDefinition(definition: DefinitionNode, printerState: PrinterState): void {
  switch (definition.kind) {
    case 'ModelDefinition':
      printModelDefinition(definition, printerState)
      break
    case 'ConfigDefinition':
      printConfigDefinition(definition, printerState)
      break
    default:
      assertNever(definition, 'Unexpected definition type')
  }
}

function printModelDefinition(definition: ModelDefinitionNode, printerState: PrinterState): void {
  const typeOffset = getTypeAlignmentOffset(definition.fields)
  printerState
    .write('model ')
    .write(definition.name.identifier)
    .write(' {')
    .newLine()
    .indent()
    .writeSeparated(definition.fields, '\n', (field) => {
      printField(field, typeOffset, printerState)
    })
    .newLine()
    .unindent()
    .write('}')
    .newLine()
}

function getTypeAlignmentOffset(fields: FieldDefinitionNode[]): number {
  const longestFieldName = Math.max(...fields.map((field) => field.name.identifier.length))
  return longestFieldName + 2 // minimum ident is 2 spaces from field name
}

function printField(field: FieldDefinitionNode, typeOffset: number, printerState: PrinterState): void {
  printerState.write(field.name.identifier).write('  ')
  printType(field.type, typeOffset, printerState)

  for (const attribute of field.attributes) {
    printerState.write(' ')
    printAttribute(attribute, printerState)
  }
}

function printType(type: TypeNode, typeOffset: number, printerState: PrinterState): void {
  printerState.writeAligned(type.name.identifier, typeOffset)
  if (type.modifier === 'array') {
    printerState.write('[]')
  } else if (type.modifier === 'optional') {
    printerState.write('?')
  }
}

function printAttribute(attribute: AttributeNode, printerState: PrinterState): void {
  printerState.write(attribute.name)
}

function printConfigDefinition(definition: ConfigDefinitionNode, printerState: PrinterState): void {
  printerState
    .write(definition.type)
    .write(' ')
    .write(definition.name.identifier)
    .write(' {')
    .newLine()
    .indent()
    .writeSeparated(definition.options, '\n', printConfigOption)
    .newLine()
    .unindent()
    .write('}')
    .newLine()
}

function printConfigOption(option: ConfigOptionNode, printerState: PrinterState): void {
  printerState.write(option.key.identifier).write(' = ')
  printExpression(option.value, printerState)
}

function printExpression(expression: ExpressionNode, printerState: PrinterState): void {
  switch (expression.kind) {
    case 'BooleanLiteral':
      printBooleanLiteral(expression, printerState)
      break
    case 'StringLiteral':
      printStringLiteral(expression, printerState)
      break
    case 'FunctionCall':
      printFunctionCall(expression, printerState)
      break
    default:
      assertNever(expression, 'Unexpected value kind')
  }
}

function printBooleanLiteral(literal: BooleanLiteralNode, printerState: PrinterState): void {
  printerState.write(String(literal.value))
}

function printStringLiteral(literal: StringLiteralNode, printerState: PrinterState): void {
  printerState.write('"').write(literal.value.replace(/"/g, '\\"')).write('"')
}

function printFunctionCall(call: FunctionCallNode, printerState: PrinterState): void {
  printerState.write(call.name.identifier).write('(').writeSeparated(call.arguments, ', ', printExpression).write(')')
}
