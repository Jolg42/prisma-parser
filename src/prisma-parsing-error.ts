import { createCodeFrame } from 'simple-code-frame'
import { Position } from './ast'

export class PrismaParsingError extends SyntaxError {
  position: Position
  source: string

  constructor(message: string, source: string, position: Position) {
    super(message)
    this.name = 'PrismaParsingError'
    this.source = source
    this.position = position
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, PrismaParsingError)
    }
  }

  toString(): string {
    // our line and column numbers are line-based, simple-code-frame ones are 0-based
    const frame = createCodeFrame(this.source, this.position.line - 1, this.position.column - 1, { colors: false })
    return `${this.name}: ${this.message}\n\n${frame}`
  }
}
