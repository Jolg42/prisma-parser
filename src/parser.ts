import { Result, ok, err } from './result'
import { Token, TokenList, TokenType } from './tokenizer'
import { DocumentNode, TypeModifier, Position, ConfigType, FunctionCallNode, ExpressionNode } from './ast'
import { PrismaParsingError } from './prisma-parsing-error'

type ParsingResult<T> = Result<T, PrismaParsingError>
type RuleFunction<ReturnType> = (state: ParserState) => ParsingResult<ReturnType>

type TupleToUnion<TupleType extends unknown[]> = TupleType[number]

export function parse(source: string): DocumentNode {
  return document(new ParserState(source)).unwrap()
}

class ParserState {
  private tokenList: TokenList
  private source: string

  constructor(source: string) {
    this.source = source
    this.tokenList = new TokenList(source)
  }

  advance(): Token {
    return this.tokenList.advance()
  }

  peek(): Token {
    return this.tokenList.peek()
  }

  getPointer(): number {
    return this.tokenList.getPointer()
  }

  restorePointer(position: number): void {
    return this.tokenList.restorePointer(position)
  }

  error<T>(message: string, position: Position): ParsingResult<T> {
    return err(new PrismaParsingError(message, this.source, position))
  }
}

const document = mapRuleResult(
  () => takeUntil(definition, 'eof'),
  ([definitions, eof]): DocumentNode => ({
    kind: 'Document' as const,
    definitions,
    location: {
      start: definitions[0]?.location.start ?? 0,
      end: eof.end,
    },
  }),
)

const definition = defineRule(() => choice('Expected config or model definition', modelDefinition, configDefinition))

const modelDefinition = mapRuleResult(
  () =>
    sequence(
      keyword('model'),
      changeErrorMessage(identifier, 'Expected model name'),
      token('{'),
      takeUntil(fieldDefinition, '}'),
    ),
  ([modelKeyword, name, , [fields, closeBrace]]) => ({
    kind: 'ModelDefinition' as const,
    name,
    fields,
    location: {
      start: modelKeyword.start,
      end: closeBrace.end,
    },
  }),
)

const fieldDefinition = mapRuleResult(
  () => sequence(changeErrorMessage(identifier, 'Expected field name'), type, zeroOrMore(attribute)),
  ([name, type, attributes]) => {
    const end = attributes.length > 0 ? attributes[attributes.length - 1].location.end : type.location.end
    return {
      kind: 'FieldDefinition' as const,
      name,
      type,
      attributes,
      location: {
        start: name.location.start,
        end,
      },
    }
  },
)

const type = mapRuleResult(
  () => sequence(changeErrorMessage(identifier, 'Expected field type'), typeModifier),
  ([name, [modifier, modifierToken]]) => {
    return {
      kind: 'Type' as const,
      name,
      modifier,
      location: {
        start: name.location.start,
        end: modifierToken ? modifierToken.end : name.location.end,
      },
    }
  },
)

const typeModifier = mapRuleResult(
  () => optional(choice('Expected type modifier', token('?'), token('[]'))),
  (token): [TypeModifier, Token | null] => {
    if (!token) {
      return ['none', null]
    }
    if (token.token === '?') {
      return ['optional', token]
    }
    return ['array', token]
  },
)

const attribute = mapRuleResult(
  () => sequence(token('attribute'), optional(argumentsList)),
  ([token, argumentsList]) => {
    const [args, endToken] = argumentsList ?? [[], token]
    return {
      kind: 'Attribute' as const,
      name: token.token,
      arguments: args,
      location: {
        start: token.start,
        end: endToken.end,
      },
    }
  },
)

const configDefinition = mapRuleResult(
  () => sequence(configType, identifier, token('{'), takeUntil(configOption, '}')),
  ([configTypeToken, name, , [options, closingBrace]]) => ({
    kind: 'ConfigDefinition' as const,
    type: configTypeToken.token as ConfigType,
    name,
    options,
    location: {
      start: configTypeToken.start,
      end: closingBrace.end,
    },
  }),
)

const configType = defineRule(() =>
  choice('Expected generator or datasource keyword', keyword('datasource'), keyword('generator')),
)

const configOption = mapRuleResult(
  () => sequence(identifier, token('='), expression),
  ([key, , value]) => ({
    kind: 'ConfigOption' as const,
    key,
    value,
    location: {
      start: key.location.start,
      end: value.location.end,
    },
  }),
)

const expression = defineRule(() => choice('Expected expression', functionCall, stringLiteral, booleanLiteral))

const stringLiteral = mapRuleResult(
  () => token('string'),
  (token) => {
    const value = token.token.replace(/^"/, '').replace(/"$/, '').replace(/\\"/g, '"')
    return {
      kind: 'StringLiteral' as const,
      value,
      location: {
        start: token.start,
        end: token.end,
      },
    }
  },
)

const booleanLiteral = mapRuleResult(
  () => choice('Expected boolean value', keyword('true'), keyword('false')),
  (keywordToken) => ({
    kind: 'BooleanLiteral' as const,
    value: keywordToken.token === 'true' ? true : false,

    location: {
      start: keywordToken.start,
      end: keywordToken.end,
    },
  }),
)

const functionCall: RuleFunction<Required<FunctionCallNode>> = mapRuleResult(
  () => sequence(identifier, argumentsList),
  ([name, [args, closeBrace]]) => ({
    kind: 'FunctionCall' as const,
    name,
    arguments: args,
    location: {
      start: name.location.start,
      end: closeBrace.end,
    },
  }),
)

const argumentsList: RuleFunction<[ExpressionNode[], Token]> = mapRuleResult(
  () => sequence(token('('), zeroOrMoreSeparated(expression, token(',')), token(')')),
  ([, args, closingBrace]) => [args, closingBrace],
)

const identifier = mapRuleResult(
  () => token('identifier'),
  (token) => ({
    kind: 'Identifier' as const,
    identifier: token.token,
    location: {
      start: token.start,
      end: token.end,
    },
  }),
)

function sequence<Elements extends unknown[]>(
  ...ruleFunctions: { [K in keyof Elements]: RuleFunction<Elements[K]> }
): RuleFunction<Elements> {
  return defineRule(() => (state) => {
    let result = ok([] as unknown[])
    for (const rule of ruleFunctions) {
      result = result.flatMap((elements) => {
        return rule(state).map((nextValue) => [...elements, nextValue])
      })
    }

    return result as ParsingResult<Elements>
  })
}

function choice<Elements extends unknown[]>(
  defaultError: string,
  ...ruleFunctions: { [K in keyof Elements]: RuleFunction<Elements[K]> }
): RuleFunction<TupleToUnion<Elements>> {
  return defineRule(() => (state) => {
    let result = state.error(defaultError, state.peek().start)
    for (const rule of ruleFunctions) {
      result = result.orElse((previousError) => {
        return rule(state).orElse((nextError) => {
          // report longest partial match as an error of choice operator
          // chances are, that if one alternative matched more tokens that rule was users intention and this error will
          // be more relevant
          const errorToReport = nextError.position.offset > previousError.position.offset ? nextError : previousError
          return err(errorToReport)
        })
      })
    }

    return result as ParsingResult<TupleToUnion<Elements>>
  })
}

function optional<ReturnType>(rule: RuleFunction<ReturnType>): RuleFunction<ReturnType | null> {
  return defineRule(() => (state) => {
    return rule(state).orElse(() => ok(null))
  })
}

function takeUntil<ReturnType>(
  elementRule: RuleFunction<ReturnType>,
  tokenType: TokenType,
): RuleFunction<[ReturnType[], Token]> {
  return defineRule(() => (state) => {
    let result: ParsingResult<ReturnType[]> = ok([])
    while (state.peek().type !== tokenType && !result.isError()) {
      result = result.flatMap((elements) => {
        return elementRule(state).map((newElement) => [...elements, newElement])
      })
    }
    return result.map((values) => [values, state.advance()])
  })
}

function zeroOrMore<ReturnType>(elementRule: RuleFunction<ReturnType>): RuleFunction<ReturnType[]> {
  return defineRule(() => (state) => {
    const result: ReturnType[] = []
    let ruleResult = elementRule(state)
    while (ruleResult.isOk()) {
      result.push(ruleResult.unwrap())
      ruleResult = elementRule(state)
    }
    return ok(result)
  })
}

function zeroOrMoreSeparated<ReturnType>(
  elementRule: RuleFunction<ReturnType>,
  separatorRule: RuleFunction<unknown>,
): RuleFunction<ReturnType[]> {
  return defineRule(() => (state) => {
    const result: ReturnType[] = []
    let ruleResult = elementRule(state)
    while (ruleResult.isOk()) {
      result.push(ruleResult.unwrap())
      ruleResult = separatorRule(state).flatMap(() => elementRule(state))
    }
    return ok(result)
  })
}

const keyword = (keywordName: string): RuleFunction<Token> => {
  return defineRule(() => (state) => {
    return identiferToken(state)
      .orElse<Token>((error) => state.error(`Expected keyword ${keywordName}`, error.position))
      .flatMap((token) => {
        if (token.token !== keywordName) {
          return state.error(`Expected keyword ${keywordName}`, token.start)
        }
        return ok(token)
      })
  })
}

const identiferToken = token('identifier')

function token(tokenType: TokenType): RuleFunction<Token> {
  return defineRule(() => (state) => {
    const token = state.advance()
    if (token.type === tokenType) {
      return ok(token)
    }
    return state.error(`Expected ${tokenType}`, token.start)
  })
}

function mapRuleResult<T, U>(ruleFactory: () => RuleFunction<T>, mapper: (result: T) => U): RuleFunction<U> {
  return defineRule(() => {
    const ruleFn = ruleFactory()
    return (state) => ruleFn(state).map(mapper)
  })
}

function changeErrorMessage<T>(ruleFn: RuleFunction<T>, message: string): RuleFunction<T> {
  return (state) => {
    return ruleFn(state).orElse((error) => state.error(message, error.position))
  }
}

function defineRule<ReturnType>(ruleFactory: () => RuleFunction<ReturnType>): RuleFunction<ReturnType> {
  let ruleFn: RuleFunction<ReturnType> | null = null
  return (state) => {
    if (!ruleFn) {
      ruleFn = ruleFactory()
    }
    const position = state.getPointer()
    const result = ruleFn(state)
    if (result.isError()) {
      state.restorePointer(position)
    }
    return result
  }
}
