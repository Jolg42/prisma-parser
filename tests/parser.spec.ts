import { parse } from '../src/parser'

test('parses empty model definition', () => {
  const doc = parse(`model Foo {}`)

  expect(doc).toMatchSnapshot()
})

test('parses multiple model definitions', () => {
  const doc = parse(`
model Foo {}
model Bar {}
`)

  expect(doc).toMatchSnapshot()
})

test('parses simple field', () => {
  const doc = parse(`
  model Foo {
      id Int
  }`)

  expect(doc).toMatchSnapshot()
})

test('parses array field', () => {
  const doc = parse(`
  model Foo {
      id Int[]
  }`)

  expect(doc).toMatchSnapshot()
})

test('parses optional field', () => {
  const doc = parse(`
  model Foo {
      id Int?
  }`)

  expect(doc).toMatchSnapshot()
})

test('parses attribute', () => {
  const doc = parse(`
  model Foo {
      id Int @id
  }`)

  expect(doc).toMatchSnapshot()
})

test('parses multiple attributes', () => {
  const doc = parse(`
  model Foo {
      id Int @id @more @attributes
  }`)

  expect(doc).toMatchSnapshot()
})

test('parses multiple fields', () => {
  const doc = parse(`
  model User {
      id Int @id
      posts Post[]
      name String
      occupation String?
  }`)

  expect(doc).toMatchSnapshot()
})

test('allows to use "model" as a field name', () => {
  const doc = parse(`
    model Car {
        model String
    }
    `)
  expect(doc).toMatchSnapshot()
})

describe('errors', () => {
  test('throws error on incomplete document', () => {
    expect(() =>
      parse(`
      model Foo {
        `),
    ).toThrow('Expected } or field definition')
  })

  test('throws error on incomplete field definition', () => {
    expect(() =>
      parse(`
      model Foo {
          name
      }
        `),
    ).toThrow('Expected field type')
  })

  test('throws error on unexpected token', () => {
    expect(() =>
      parse(`
      model Foo [
          name Int
      ]
        `),
    ).toThrow('Unexpected token')
  })

  test('throws error on missing {', () => {
    expect(() =>
      parse(`
      model Foo @attr {
        name Int
      }
        `),
    ).toThrow('Expected {')
  })

  test('throws error on missing model name', () => {
    expect(() =>
      parse(`
      model {
        name Int
      }
        `),
    ).toThrow('Expected model name')
  })
})
