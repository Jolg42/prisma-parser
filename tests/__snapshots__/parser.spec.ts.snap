// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errors throws error on incomplete config definition 1`] = `
"PrismaParsingError: Expected identifier

  1 | 
  2 |       generator client {
> 3 |         
    |         ^
"
`;

exports[`errors throws error on incomplete document 1`] = `
"PrismaParsingError: Expected field name

  1 | 
  2 |       model Foo {
> 3 |         
    |         ^
"
`;

exports[`errors throws error on incomplete field definition 1`] = `
"PrismaParsingError: Expected field type

  2 |       model Foo {
  3 |           name
> 4 |       }
    |       ^
  5 |         
"
`;

exports[`errors throws error on missing { 1`] = `
"PrismaParsingError: Expected {

  1 | 
> 2 |       model Foo @attr {
    |                 ^
  3 |         name Int
  4 |       }
  5 |         
"
`;

exports[`errors throws error on missing config value 1`] = `
"PrismaParsingError: Expected expression

  2 |       datasource db1 {
  3 |         option =
> 4 |       }
    |       ^
  5 |         
"
`;

exports[`errors throws error on missing model name 1`] = `
"PrismaParsingError: Expected model name

  1 | 
> 2 |       model {
    |             ^
  3 |         name Int
  4 |       }
  5 |         
"
`;

exports[`errors throws error on unexpected token 1`] = `
"PrismaParsingError: Unexpected token

  1 | 
> 2 |       model Foo [
    |                 ^
  3 |           name Int
  4 |       ]
  5 |         
"
`;

exports[`parses prisma example schema 1`] = `
Object {
  "definitions": Array [
    Object {
      "fields": Array [
        Object {
          "attributes": Array [
            Object {
              "kind": "Attribute",
              "location": Object {
                "end": Object {
                  "column": 17,
                  "line": 2,
                  "offset": 29,
                },
                "start": Object {
                  "column": 14,
                  "line": 2,
                  "offset": 26,
                },
              },
              "name": "@id",
            },
          ],
          "kind": "FieldDefinition",
          "location": Object {
            "end": Object {
              "column": 17,
              "line": 2,
              "offset": 29,
            },
            "start": Object {
              "column": 3,
              "line": 2,
              "offset": 15,
            },
          },
          "name": Object {
            "identifier": "id",
            "kind": "Identifier",
            "location": Object {
              "end": Object {
                "column": 5,
                "line": 2,
                "offset": 17,
              },
              "start": Object {
                "column": 3,
                "line": 2,
                "offset": 15,
              },
            },
          },
          "type": Object {
            "kind": "Type",
            "location": Object {
              "end": Object {
                "column": 13,
                "line": 2,
                "offset": 25,
              },
              "start": Object {
                "column": 10,
                "line": 2,
                "offset": 22,
              },
            },
            "modifier": "none",
            "name": Object {
              "identifier": "Int",
              "kind": "Identifier",
              "location": Object {
                "end": Object {
                  "column": 13,
                  "line": 2,
                  "offset": 25,
                },
                "start": Object {
                  "column": 10,
                  "line": 2,
                  "offset": 22,
                },
              },
            },
          },
        },
        Object {
          "attributes": Array [],
          "kind": "FieldDefinition",
          "location": Object {
            "end": Object {
              "column": 16,
              "line": 3,
              "offset": 45,
            },
            "start": Object {
              "column": 3,
              "line": 3,
              "offset": 32,
            },
          },
          "name": Object {
            "identifier": "name",
            "kind": "Identifier",
            "location": Object {
              "end": Object {
                "column": 7,
                "line": 3,
                "offset": 36,
              },
              "start": Object {
                "column": 3,
                "line": 3,
                "offset": 32,
              },
            },
          },
          "type": Object {
            "kind": "Type",
            "location": Object {
              "end": Object {
                "column": 16,
                "line": 3,
                "offset": 45,
              },
              "start": Object {
                "column": 10,
                "line": 3,
                "offset": 39,
              },
            },
            "modifier": "none",
            "name": Object {
              "identifier": "String",
              "kind": "Identifier",
              "location": Object {
                "end": Object {
                  "column": 16,
                  "line": 3,
                  "offset": 45,
                },
                "start": Object {
                  "column": 10,
                  "line": 3,
                  "offset": 39,
                },
              },
            },
          },
        },
        Object {
          "attributes": Array [],
          "kind": "FieldDefinition",
          "location": Object {
            "end": Object {
              "column": 16,
              "line": 4,
              "offset": 61,
            },
            "start": Object {
              "column": 3,
              "line": 4,
              "offset": 48,
            },
          },
          "name": Object {
            "identifier": "email",
            "kind": "Identifier",
            "location": Object {
              "end": Object {
                "column": 8,
                "line": 4,
                "offset": 53,
              },
              "start": Object {
                "column": 3,
                "line": 4,
                "offset": 48,
              },
            },
          },
          "type": Object {
            "kind": "Type",
            "location": Object {
              "end": Object {
                "column": 16,
                "line": 4,
                "offset": 61,
              },
              "start": Object {
                "column": 10,
                "line": 4,
                "offset": 55,
              },
            },
            "modifier": "none",
            "name": Object {
              "identifier": "String",
              "kind": "Identifier",
              "location": Object {
                "end": Object {
                  "column": 16,
                  "line": 4,
                  "offset": 61,
                },
                "start": Object {
                  "column": 10,
                  "line": 4,
                  "offset": 55,
                },
              },
            },
          },
        },
        Object {
          "attributes": Array [],
          "kind": "FieldDefinition",
          "location": Object {
            "end": Object {
              "column": 14,
              "line": 5,
              "offset": 75,
            },
            "start": Object {
              "column": 3,
              "line": 5,
              "offset": 64,
            },
          },
          "name": Object {
            "identifier": "age",
            "kind": "Identifier",
            "location": Object {
              "end": Object {
                "column": 6,
                "line": 5,
                "offset": 67,
              },
              "start": Object {
                "column": 3,
                "line": 5,
                "offset": 64,
              },
            },
          },
          "type": Object {
            "kind": "Type",
            "location": Object {
              "end": Object {
                "column": 14,
                "line": 5,
                "offset": 75,
              },
              "start": Object {
                "column": 10,
                "line": 5,
                "offset": 71,
              },
            },
            "modifier": "optional",
            "name": Object {
              "identifier": "Int",
              "kind": "Identifier",
              "location": Object {
                "end": Object {
                  "column": 13,
                  "line": 5,
                  "offset": 74,
                },
                "start": Object {
                  "column": 10,
                  "line": 5,
                  "offset": 71,
                },
              },
            },
          },
        },
        Object {
          "attributes": Array [],
          "kind": "FieldDefinition",
          "location": Object {
            "end": Object {
              "column": 16,
              "line": 6,
              "offset": 91,
            },
            "start": Object {
              "column": 3,
              "line": 6,
              "offset": 78,
            },
          },
          "name": Object {
            "identifier": "posts",
            "kind": "Identifier",
            "location": Object {
              "end": Object {
                "column": 8,
                "line": 6,
                "offset": 83,
              },
              "start": Object {
                "column": 3,
                "line": 6,
                "offset": 78,
              },
            },
          },
          "type": Object {
            "kind": "Type",
            "location": Object {
              "end": Object {
                "column": 16,
                "line": 6,
                "offset": 91,
              },
              "start": Object {
                "column": 10,
                "line": 6,
                "offset": 85,
              },
            },
            "modifier": "array",
            "name": Object {
              "identifier": "Post",
              "kind": "Identifier",
              "location": Object {
                "end": Object {
                  "column": 14,
                  "line": 6,
                  "offset": 89,
                },
                "start": Object {
                  "column": 10,
                  "line": 6,
                  "offset": 85,
                },
              },
            },
          },
        },
      ],
      "kind": "ModelDefinition",
      "location": Object {
        "end": Object {
          "column": 2,
          "line": 7,
          "offset": 93,
        },
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "name": Object {
        "identifier": "User",
        "kind": "Identifier",
        "location": Object {
          "end": Object {
            "column": 11,
            "line": 1,
            "offset": 10,
          },
          "start": Object {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
        },
      },
    },
    Object {
      "fields": Array [
        Object {
          "attributes": Array [
            Object {
              "kind": "Attribute",
              "location": Object {
                "end": Object {
                  "column": 19,
                  "line": 10,
                  "offset": 126,
                },
                "start": Object {
                  "column": 16,
                  "line": 10,
                  "offset": 123,
                },
              },
              "name": "@id",
            },
          ],
          "kind": "FieldDefinition",
          "location": Object {
            "end": Object {
              "column": 19,
              "line": 10,
              "offset": 126,
            },
            "start": Object {
              "column": 3,
              "line": 10,
              "offset": 110,
            },
          },
          "name": Object {
            "identifier": "id",
            "kind": "Identifier",
            "location": Object {
              "end": Object {
                "column": 5,
                "line": 10,
                "offset": 112,
              },
              "start": Object {
                "column": 3,
                "line": 10,
                "offset": 110,
              },
            },
          },
          "type": Object {
            "kind": "Type",
            "location": Object {
              "end": Object {
                "column": 15,
                "line": 10,
                "offset": 122,
              },
              "start": Object {
                "column": 12,
                "line": 10,
                "offset": 119,
              },
            },
            "modifier": "none",
            "name": Object {
              "identifier": "Int",
              "kind": "Identifier",
              "location": Object {
                "end": Object {
                  "column": 15,
                  "line": 10,
                  "offset": 122,
                },
                "start": Object {
                  "column": 12,
                  "line": 10,
                  "offset": 119,
                },
              },
            },
          },
        },
        Object {
          "attributes": Array [],
          "kind": "FieldDefinition",
          "location": Object {
            "end": Object {
              "column": 18,
              "line": 11,
              "offset": 144,
            },
            "start": Object {
              "column": 3,
              "line": 11,
              "offset": 129,
            },
          },
          "name": Object {
            "identifier": "title",
            "kind": "Identifier",
            "location": Object {
              "end": Object {
                "column": 8,
                "line": 11,
                "offset": 134,
              },
              "start": Object {
                "column": 3,
                "line": 11,
                "offset": 129,
              },
            },
          },
          "type": Object {
            "kind": "Type",
            "location": Object {
              "end": Object {
                "column": 18,
                "line": 11,
                "offset": 144,
              },
              "start": Object {
                "column": 12,
                "line": 11,
                "offset": 138,
              },
            },
            "modifier": "none",
            "name": Object {
              "identifier": "String",
              "kind": "Identifier",
              "location": Object {
                "end": Object {
                  "column": 18,
                  "line": 11,
                  "offset": 144,
                },
                "start": Object {
                  "column": 12,
                  "line": 11,
                  "offset": 138,
                },
              },
            },
          },
        },
        Object {
          "attributes": Array [],
          "kind": "FieldDefinition",
          "location": Object {
            "end": Object {
              "column": 18,
              "line": 12,
              "offset": 162,
            },
            "start": Object {
              "column": 3,
              "line": 12,
              "offset": 147,
            },
          },
          "name": Object {
            "identifier": "content",
            "kind": "Identifier",
            "location": Object {
              "end": Object {
                "column": 10,
                "line": 12,
                "offset": 154,
              },
              "start": Object {
                "column": 3,
                "line": 12,
                "offset": 147,
              },
            },
          },
          "type": Object {
            "kind": "Type",
            "location": Object {
              "end": Object {
                "column": 18,
                "line": 12,
                "offset": 162,
              },
              "start": Object {
                "column": 12,
                "line": 12,
                "offset": 156,
              },
            },
            "modifier": "none",
            "name": Object {
              "identifier": "String",
              "kind": "Identifier",
              "location": Object {
                "end": Object {
                  "column": 18,
                  "line": 12,
                  "offset": 162,
                },
                "start": Object {
                  "column": 12,
                  "line": 12,
                  "offset": 156,
                },
              },
            },
          },
        },
        Object {
          "attributes": Array [],
          "kind": "FieldDefinition",
          "location": Object {
            "end": Object {
              "column": 16,
              "line": 13,
              "offset": 178,
            },
            "start": Object {
              "column": 3,
              "line": 13,
              "offset": 165,
            },
          },
          "name": Object {
            "identifier": "author",
            "kind": "Identifier",
            "location": Object {
              "end": Object {
                "column": 9,
                "line": 13,
                "offset": 171,
              },
              "start": Object {
                "column": 3,
                "line": 13,
                "offset": 165,
              },
            },
          },
          "type": Object {
            "kind": "Type",
            "location": Object {
              "end": Object {
                "column": 16,
                "line": 13,
                "offset": 178,
              },
              "start": Object {
                "column": 12,
                "line": 13,
                "offset": 174,
              },
            },
            "modifier": "none",
            "name": Object {
              "identifier": "User",
              "kind": "Identifier",
              "location": Object {
                "end": Object {
                  "column": 16,
                  "line": 13,
                  "offset": 178,
                },
                "start": Object {
                  "column": 12,
                  "line": 13,
                  "offset": 174,
                },
              },
            },
          },
        },
      ],
      "kind": "ModelDefinition",
      "location": Object {
        "end": Object {
          "column": 2,
          "line": 14,
          "offset": 180,
        },
        "start": Object {
          "column": 1,
          "line": 9,
          "offset": 95,
        },
      },
      "name": Object {
        "identifier": "Post",
        "kind": "Identifier",
        "location": Object {
          "end": Object {
            "column": 11,
            "line": 9,
            "offset": 105,
          },
          "start": Object {
            "column": 7,
            "line": 9,
            "offset": 101,
          },
        },
      },
    },
  ],
  "kind": "Document",
  "location": Object {
    "end": Object {
      "column": 1,
      "line": 15,
      "offset": 181,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
}
`;
